### ISSUE TITLE
Very slow on big files.

### ISSUE THREAD

**Issue by trusktr** [VSCodeVim/Vim#3924]
**Describe the bug**

All Vim shortcuts and keybindings grind to a halt when inside of a very large file.

However, normal editing mode (turning off the Bim plugin to restore all normal editing features of VS Code), works fine, and I can navigate around large files just fine.

**To Reproduce**

Open a huge code file, for example with 280,000 lines of code.

**Expected behavior**

Editing with Vim plugin turned on should be as fast as non-Vim regular mode (plugin turned off).

**Screenshots**
If remapping-related, please attach log output: https://github.com/VSCodeVim/Vim#debugging-remappings.

**Environment (please complete the following information):**

- Extension (VsCodeVim) version: 1.8.2
- VSCode version: 1.36.1
- OS: macOS

**Additional context**

Is plugin integration in Atom more flexible? In Atom, the vim-mode-plus plugin seems to integrate very well: it scripts all of Atom's features, rather than writing new parallel features.

For example, when I undo in Atom vim mode, the dot in the tab of the file (the dot that tells you when the file is modified) goes away when we undo back to the saved state. But in VS Code Vim mode, the dot never goes away, even if we undo to the state we know is save. 

That, along with https://github.com/VSCodeVim/Vim/issues/3871, indicates that VS Code Vim implements features in parallel rather than scripting existing features.

My general impression about VS Code is that writing plugins is a bit more difficult in VS Code because plugin API is more limited (which makes sense, and this allows VS Code's Live Share feature to be much more advanced than Atom's equivalent, because VS Code can enforce a higher level of consistency across editor features due to a more restricted plugin API).

Is VS Code's more restricted plugin API a reason why some features seem implemented in parallel, because otherwise they aren't possible to script?

**Comment by bdwiel** [comment]
I'm having this issue, too. My files aren't quite as big as 280k, but more like 10k. Working with large files used to not be noticeable with the plugin until sometime in the past 2-3 months.

**Environment (please complete the following information):**

* Extension (VsCodeVim) version: 1.9.0
* VSCode version: 1.36.1
* OS: macOS

**Comment by bdwiel** [comment]
I forgot to mention, the performance hit is there even if I am working on a <1k line file but have a large file open in another workspace and window.

**Comment by mheiber** [comment]
Thanks for looking into this issue. Fwiw, slowness would be more tolerable if the extension lagged, but kept an accurate buffer. Unfortunately, when I type quickly in a large file the order of keystrokes is messed up, making the extension unusable for large files. A good test example is the 33KLOC src/compiler/checker.ts in the Microsoft/TypeScript codebase. I've seen the ordering issue in smaller files, but the issue happens in checker.ts every time.

**Comment by vlad2135** [comment]
Seems related to #2216

**Comment by J-Fields** [comment]
Version 1.13.0 introduces the `vim.experimentalOptimizations` setting, which significantly improves the performance in large files.

**Comment by goodloop** [comment]
> Version 1.13.0 introduces the `vim.experimentalOptimizations` setting, which significantly improves the performance in large files.

works for me with 5k lines file( maybe I should refactory my code now )

**Comment by J-Fields** [comment]
Version 1.14.0 will also improve on this front. The problem is mainly caused by our history tracking for undo/redo, which requires #1490 to resolve completely.

**Comment by guledali** [comment]
@J-Fields Why did you close this one? I was about to open an issue on this

**Comment by J-Fields** [comment]
@guledali How large is the file you're working in, what language (I assume it's code), and what other extensions do you have installed?

**Comment by guledali** [comment]
Quite large it's a TypeScript codebase, other extensions  installed is 

[Apollo GraphQL](https://marketplace.visualstudio.com/items?itemName=apollographql.vscode-apollo)

[Create Files & Folders : On The Go](https://marketplace.visualstudio.com/items?itemName=ritwickdey.create-file-folder)

[ESLint](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint)

[File Utils](https://marketplace.visualstudio.com/items?itemName=sleistner.vscode-fileutils)


[Git Automator](https://marketplace.visualstudio.com/items?itemName=ivangabriele.vscode-git-add-and-commit)

[Git Graph](https://marketplace.visualstudio.com/items?itemName=mhutchie.git-graph)


[GitLens](https://marketplace.visualstudio.com/items?itemName=eamodio.gitlens)


Gruvbox theme & material icons

Polacode

Prettier

Settings sync

Vim

**Comment by J-Fields** [comment]
Can you give me a rough line count?

**Comment by guledali** [comment]
It's laggy takes couple second for example `dd` does not delete a line directly.

`i` I have to press it couple times in order to get to insert mode. 

Undo `u` is the slowest of all, I don't know how to explain but it's laggy as hell

**Comment by guledali** [comment]
@J-Fields Do you have fix for this?

**Comment by J-Fields** [comment]
@guledali The short answer is no, not really. You're welcome to open a new issue and we can continue digging into this, but unless another extension is acting up (the ones you listed don't seem problematic to me), there's probably not much that can be done in the short term. I'm a bit surprised you're having drastic issues on files that are big but not HUGE, though we'll never compete with native Vim in speed.

**Comment by listenerri** [comment]
vscode disables the extensions from accessing the current editor object (vscode.window.activeTextEditor), which causes the vim extension to not work properly

I created a related issue in the vscode repository: https://github.com/microsoft/vscode/issues/276277

That issue requires at least 20 upvotes to be addressed, if you feel this should be fixed/implemented, please express your thoughts in that issue
